File: solutions/javascript/0664-strange-printer.js
/**
 * @param {string} s
 * @return {number}
 */
var strangePrinter = function(s) {
    let n = s.length;
    let dp = Array.from(Array(n), () => new Array(n).fill(0));
    
    for (let i = n-1; i >= 0; --i) {
        dp[i][i] = 1;
        for (let j = i+1; j < n; ++j) {
            dp[i][j] = dp[i][j-1] + 1;
            for (let k = i; k < j; ++k) {
                if (s[k] == s[j]) {
                    dp[i][j] = Math.min(dp[i][j], dp[i][k] + (k+1<=j-1 ? dp[k+1][j-1] : 0));
                }
            }
        }
    }
    return dp[0][n-1];
};

File: solutions/ruby/0664-strange-printer.rb
def strange_printer(s)
    n = s.size
    dp = Array.new(n) {Array.new(n, 0)}

    (n-1).downto(0) do |i|
        dp[i][i] = 1
        (i+1...n).each do |j|
            dp[i][j] = dp[i][j-1] + 1
            (i...j).each do |k|
                if s[k] == s[j]
                    dp[i][j] = [dp[i][j], dp[i][k] + (k+1<=j-1 ? dp[k+1][j-1] : 0)].min
                end
            end
        end
    end

    dp[0][n-1]
end
